pkg/cri/resource-manager/visualizer/visualizer.go

```diff
@@ -0,0 +1,213 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package visualizer
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	xhttp "github.com/intel/cri-resource-manager/pkg/instrumentation/http"
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+)
+
+const (
+	// HTTP URI prefix to register all visualizer implementations under.
+	visualizerPrefix = "/ui"
+)
+
+// Our logger instance.
+var log = logger.NewLogger("visualizer")
+
+// visualizer captures our runtime state.
+type visualizer struct {
+	sync.Mutex
+	internal map[string]http.FileSystem
+	external map[string]http.FileSystem
+}
+
+var visualizers = &visualizer{
+	internal: map[string]http.FileSystem{},
+	external: map[string]http.FileSystem{},
+}
+
+// RegisterInternal registers an internal visualizer implementation.
+func RegisterInternal(name string, dir http.FileSystem) {
+	visualizers.register(name, dir, true)
+}
+
+// RegisterExternal registers an external visualizer implementation.
+func RegisterExternal(name string, dir http.FileSystem) {
+	visualizers.register(name, dir, false)
+}
+
+// Setup sets up the given multiplexer to serve visualization implementations.
+func Setup(mux *xhttp.ServeMux) error {
+	log.Info("activating visualization interface...")
+
+	mux.Handle("/", http.RedirectHandler("/ui/index.html", http.StatusFound))
+	mux.Handle("/ui", http.RedirectHandler("/ui/index.html", http.StatusFound))
+	mux.Handle("/ui/internal/", http.FileServer(visualizers))
+	mux.Handle("/ui/external/", http.FileServer(visualizers))
+	mux.HandleFunc("/ui/index.html", visualizers.genIndexHTML)
+
+	return nil
+}
+
+// Open is the http.Dir implementation for our visualizers.
+func (v *visualizer) Open(path string) (http.File, error) {
+	var dir http.FileSystem
+	var ok bool
+
+	log.Debug("HTTP request for '%s'", path)
+
+	rpath, err := filepath.Rel(visualizerPrefix+"/", path)
+	if err != nil {
+		log.Error("failed to resolve path for %s: %v", path, err)
+		return nil, visualizerError("failed to resolve path for %s: %v", path, err)
+	}
+
+	log.Debug("%s => %s", path, rpath)
+
+	split := strings.Split(rpath, "/")
+	if len(split) < 1 {
+		return nil, visualizerError("failed to resolve relative path %s", rpath)
+	}
+
+	kind := split[0]
+
+	switch {
+	case len(split) > 1 && kind == "internal":
+		dir, ok = v.internal[split[1]]
+	case len(split) > 1 && kind == "external":
+		dir, ok = v.external[split[1]]
+	default:
+		ok = false
+	}
+
```

Comment:
Could be simplified with smth like:
```
	ok = false
	if len(split) > 1 {
		kind := split[0]
		switch kind {
		case "internal":
			dir, ok = v.internal[split[1]]
		case "external":
			dir, ok = v.external[split[1]]
		}
	}
```
WDYT?

Reply:
True. Updated.